### PR TITLE
Feature/Auto-App-Sync: App Registration

### PR DESCRIPTION
**Pull request by krystalleon**
# Description

Added backend functions: Application & ApiKey database tables, repositories, services, and application controller.
Added frontend view, AppRegister.vue

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

I don't have any unit tests yet, but I ran it locally and tried submitting different field scenarios through the front end:
-blank fields
-invalid email
-invalid urls
-already existing application
Then I ran queries to check the database contents.

# Notes

-In AppRegister.vue, the POST url is currently set to localhost, so that will eventually need to be changed to the SSO url.
-I added the app registration view to the navigation bar, but let me know if the app-related pages should be organized differently.

### CODE REVIEW

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (alfredovargasjr) [inline Backend/ServiceLayer/Services/ApiKeyService.cs]: Maybe a more secure password can be a hex value generated by using the GUID class and having the dashes removed?

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (alfredovargasjr) [inline Backend/ServiceLayer/Services/ApiKeyService.cs]: what currently isn't working for this?

File: Backend/DataAccessLayer/Models/ApiKey.cs
Comment (alfredovargasjr) [inline Backend/DataAccessLayer/Models/ApiKey.cs:18]: The primary key should be a unique identifier such as a GUID to keep consistency with the other models in the database.

File: Backend/DataAccessLayer/Models/Application.cs
Comment (alfredovargasjr) [inline Backend/DataAccessLayer/Models/Application.cs:19]: Same thing regarding the primary key being a GUID.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (alfredovargasjr) [inline Backend/WebAPI/Controllers/ApplicationController.cs]: A manager class can be used to move all of this logic into its own abstraction layer. This can make the router cleaner and smaller, but will add a bit of redundancy. The response messages can be returned from the manager class

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (krystalleon) [reply Backend/ServiceLayer/Services/ApiKeyService.cs]: Ok. I will look into that.

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (krystalleon) [reply Backend/ServiceLayer/Services/ApiKeyService.cs]: I put together some code, but I haven't tested it yet, so it's a placeholder to work on in the future.

File: Backend/DataAccessLayer/Models/ApiKey.cs
Comment (krystalleon) [reply Backend/DataAccessLayer/Models/ApiKey.cs:18]: Ok.  I will add the GUID key to both ApiKey and Application models

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (krystalleon) [reply Backend/WebAPI/Controllers/ApplicationController.cs]: Got it.  I'll reorganize.

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Maxlength of 128 might be a bit short here. Also, might want to call this "LaunchUrl" since your other url is "UserDeletionUrl"

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Rename this to LogoUrl

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Needs a UUID as a surrogate key "ID"

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Needs a UUID as a surrogate key "ID"

Comment (julianpoy) [reply Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Also needs a reference to the surrogate key of the application it belongs to "ApplicationID"

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: These two lines should not be changed here.

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Remove these statements

File: Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903060321294_create-tables-applications-apikeys.cs]: Remove this table definition - We don't have a claims system in the SSO

File: Backend/DataAccessLayer/Repositories/ApplicationRepository.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Repositories/ApplicationRepository.cs]: Probably not the best idea to reference this by URL. Use the surrogate UUID here.

File: Backend/DataAccessLayer/Repositories/ApplicationRepository.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Repositories/ApplicationRepository.cs]: Again, probably best not to find an application by it's launch URL.

File: Backend/DataAccessLayer/Repositories/ApplicationRepository.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Repositories/ApplicationRepository.cs]: I don't think this function provides much value after switching to a UUID system.

File: Backend/Dev/Dev.csproj
Comment (julianpoy) [inline Backend/Dev/Dev.csproj]: What is the reason for this removal?

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (julianpoy) [inline Backend/ServiceLayer/Services/ApiKeyService.cs]: 👍

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (julianpoy) [reply Backend/ServiceLayer/Services/ApiKeyService.cs]: As I mentioned to @WinnMoo - the Session service/repo has a very good crypto random generator. Please reuse that by making that a reusable module shared by the session service (as well as his password reset functionality)

File: Backend/ServiceLayer/Services/ApiKeyService.cs
Comment (julianpoy) [reply Backend/ServiceLayer/Services/ApiKeyService.cs]: Please make sure to reuse/share the code that @WinnMoo is writing for password reset.

File: Backend/ServiceLayer/Services/ApplicationService.cs
Comment (julianpoy) [inline Backend/ServiceLayer/Services/ApplicationService.cs]: This isn't really very useful. Every request should create a new application.

File: Backend/WebAPI/App_Start/WebApiConfig.cs
Comment (julianpoy) [inline Backend/WebAPI/App_Start/WebApiConfig.cs]: [non-blocking] What is this for? (curiosity)

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (julianpoy) [inline Backend/WebAPI/Controllers/ApplicationController.cs]: [non-blocking] Consider renaming this route to "api/applications/create"  - "register" terminology may be confused with user registration.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (julianpoy) [inline Backend/WebAPI/Controllers/ApplicationController.cs]: This really should never be a case. I'd remove this.

If you leave it like this, then I could claim your application URL and you'd never be able to create an entry for your app.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (julianpoy) [inline Backend/WebAPI/Controllers/ApplicationController.cs]: This shouldn't be a 400 level error - the system failed to save the changes to the database. This should probably be a 500 level error.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (julianpoy) [inline Backend/WebAPI/Controllers/ApplicationController.cs]: Are we accepting multipart data? Remember, just store an image URL.

**Review by julianpoy — DISMISSED** [review]

File: Frontend/src/main.js
Comment (julianpoy) [inline Frontend/src/main.js:26]: Remove this placeholder code before merge

File: Frontend/src/router.js
Comment (julianpoy) [inline Frontend/src/router.js:36]: [non-blocking] Can we name this something other than app-register? I feel like this route is confusing.

File: Frontend/src/views/AppRegister.vue
Comment (julianpoy) [inline Frontend/src/views/AppRegister.vue]: Remove the variable "hide"

Instead, just v-if="key"

File: Frontend/src/views/AppRegister.vue
Comment (julianpoy) [inline Frontend/src/views/AppRegister.vue:3]: Instead of the "disabled" variable, just v-if="!key" this form

File: Backend/Dev/Dev.csproj
Comment (krystalleon) [reply Backend/Dev/Dev.csproj]: Unit Testing was giving me errors before I pulled the latest version.  I forgot to add it back.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (krystalleon) [reply Backend/WebAPI/Controllers/ApplicationController.cs]: No.  The Publish code is outdated from before our conversation.  I still have to make those changes.

File: Backend/WebAPI/App_Start/WebApiConfig.cs
Comment (krystalleon) [reply Backend/WebAPI/App_Start/WebApiConfig.cs]: As we discussed, we are now using just ImageUrl instead of the image file, so this line is no longer necessary.  I will delete it.

File: Backend/WebAPI/Controllers/ApplicationController.cs
Comment (julianpoy) [reply Backend/WebAPI/Controllers/ApplicationController.cs]: :+1:

File: Backend/ServiceLayer/Services/ApplicationService.cs
Comment (krystalleon) [reply Backend/ServiceLayer/Services/ApplicationService.cs]: What if you try to register with the same information multiple times?  The Guid generates a new id every time, so the applications would not be considered "the same".  I understand what you mentioned below that someone could claim your app url if you check for it existing.  So is it ok to have multiple entries leading to the same application?

**Comment by alfredovargasjr** [comment]
Resolve conflicts and accept the changes from master, the old web-api project will be deleted so you will have to move the logic onto the new web-api project. The old one had too much bloat.

**Comment by krystalleon** [comment]
I merged master into my branch ok.  The only problem it has is my Application controller is still blocked by cors unless I have the [EnabledCors("*", "*", "*)] over my controller class

File: Backend/KFC-WebAPI/Content/bootstrap-grid.css
Comment (alfredovargasjr) [inline Backend/KFC-WebAPI/Content/bootstrap-grid.css]: KFC-WebAPI/content should be deleted. It has no purpose.

File: Backend/KFC-WebAPI/Scripts/bootstrap.bundle.js
Comment (alfredovargasjr) [inline Backend/KFC-WebAPI/Scripts/bootstrap.bundle.js]: KFC-WebAPI/Scripts should be deleted. This has no purpose

**Review by alfredovargasjr — CHANGES_REQUESTED** [review]

**Review by alfredovargasjr — APPROVED** [review]

**Review by alfredovargasjr — DISMISSED** [review]

File: Backend/DataAccessLayer/Migrations/201903070039162_create-tables-applications-apikeys.cs
Comment (julianpoy) [inline Backend/DataAccessLayer/Migrations/201903070039162_create-tables-applications-apikeys.cs:41]: I believe you need to drop the PrimaryKey reference in your down() here.

**Review by julianpoy — APPROVED** [review]

**Pull request merged**
